### DISCUSSION TITLE
found something weird when installing kiali using helm

### DISCUSSION THREAD

**Discussion by lwabish** [kiali/kiali#5398] [General]
I am including kiali/kiali-operator chart as a dependency of my own chart, but after installing the parent chart, the Kiali CRD is not installed by helm, although the crds directory exists.

Chart.yaml:
```
...
dependencies:
  - repository: "alias:kiali"
    name: kiali-operator
    version: 1.55.0
...
```

` helm upgrade -i -n istio-system --create-namespace istio ./`

> Error: UPGRADE FAILED: resource mapping not found for name: "kiali" namespace: "istio-system" from "": no matches for kind "Kiali" in version "kiali.io/v1alpha1"  ensure CRDs are installed first

Also ,last week I run into another similar situation somehow:  kiali operator is installed, but the pod failed because of the missing of CRD Kiali.

Then I tried install kiali/kiali-operator directly, not as a dependency of a parent chart, and the crd is installed by helm sucessfully.
`helm upgrade -i -n istio-system --create-namespace kiali kiali/kiali-operator --version 1.55.0`

I wonder what's the situation. I have some guesses:
- `helm upgrade -i --create-namespace` is not fully equivalent to `helm install` when a CRD exists?
-  the CRD is included after the chart has been installed, maybe a CRD must be included at the very first install of the whole chart?

**Comment by jshaughn** [comment]
cc @jmazzitelli

**Comment by jmazzitelli** [comment]
CRD management is pretty much "hands-off" in Helm. In earlier versions of Helm, they did try to manage CRDs, but found it too difficult due to the nature of kubernetes CRDs themselves. Please see [helm docs here](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations) where you read:

```
There is no support at this time for upgrading or deleting CRDs using Helm.
This was an explicit decision after much community discussion due to the
danger for unintentional data loss. Furthermore, there is currently no community
consensus around how to handle CRDs and their lifecycle. 
```
Helm does have the ability to install CRDs (from the crds directory in the chart - see [Helm docs here](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you)) but that is about it. It won't update or manage them, as the Helm docs above indicate.

You must install the CRD at Kiali operator install time (`helm install`). If there is no CRD, then `helm update` will most likely fail.

**Comment by jmazzitelli** [comment]
BTW: your question is probably best directed to the Helm community. But I bet if you search the Helm forums, you will find tons of information/threads on CRD management and how to get Helm to not only install CRDs but then to wait for CRDs to be established before continuing to install/update the templated resources. Your situation has the added wrinkle of using dependencies - again, the Helm community would probably be of help for that also.

**Reply by lwabish** [reply]
Thanks a lot for your reply ,  learned a lot from that!
